src/diagnostics.rs

```diff
@@ -28,16 +31,21 @@ pub struct DiagnosticsConfig {
 
 #[derive(Debug, Clone)]
 pub struct Diagnostic {
-  message: String,
-  severity: Severity,
+  pub message: String,
+  pub severity: Severity,
+  pub span: Option<FileSpan>,
 }
 
 #[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
 pub enum DiagnosticOrigin {
+  /// An error when parsing source code.
+  Parsing,
   /// An error from the relationship between multiple top-level definitions.
   Book,
   /// An error in a pattern-matching function definition rule.
   Rule(Name),
+  /// An error when transforming `imp` syntax into `fun`.
+  Function(Name),
```

Comment:
`Function` and `Rule` diagnostics are the same, you could have just a single one called `Function`.

I don't think its very meaningful which syntax it comes from.